File: compat/suites/java-sdk/src/main/java/io/overcast/compat/groups/KmsGroup.java
package io.overcast.compat.groups;

import io.overcast.compat.clients.AwsClients;
import io.overcast.compat.harness.Assertions;
import io.overcast.compat.harness.TestContext;
import io.overcast.compat.harness.TestFn;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.services.kms.KmsClient;
import software.amazon.awssdk.services.kms.model.*;

import java.util.Map;

/**
 * KMS compatibility test group.
 *
 * <p>Groups: kms-keys, kms-crypto, kms-asymmetric.
 */
public final class KmsGroup implements ServiceGroup {

    private final AwsClients clients;

    public KmsGroup(AwsClients clients) {
        this.clients = clients;
    }

    private KmsClient kms() { return clients.kms(); }

    @Override
    public Map<String, TestFn> impls() {
        return Map.ofEntries(
                Map.entry("CreateKey",                      this::createKey),
                Map.entry("DescribeKey",                    this::describeKey),
                Map.entry("CreateKmsAlias",                 this::createKmsAlias),
                Map.entry("ListKmsAliases",                 this::listKmsAliases),
                Map.entry("ListKeys",                       this::listKeys),
                Map.entry("DisableKey",                     this::disableKey),
                Map.entry("EnableKey",                      this::enableKey),
                Map.entry("ScheduleKeyDeletion",            this::scheduleKeyDeletion),
                Map.entry("CancelKeyDeletion",              this::cancelKeyDeletion),
                Map.entry("TagKMSResource",                 this::tagKmsResource),
                Map.entry("UntagKMSResource",               this::untagKmsResource),
                Map.entry("ListKMSResourceTags",            this::listKmsResourceTags),
                Map.entry("Encrypt",                        this::encrypt),
                Map.entry("Decrypt",                        this::decrypt),
                Map.entry("GenerateDataKey",                this::generateDataKey),
                Map.entry("GenerateDataKeyWithoutPlaintext",this::generateDataKeyWithoutPlaintext),
                Map.entry("Sign",                           this::sign),
                Map.entry("Verify",                         this::verify)
        );
    }

    @Override
    public Map<String, TestFn> setups() {
        return Map.ofEntries(
                Map.entry("kms-keys",       this::setupKeys),
                Map.entry("kms-crypto",     this::setupCrypto),
                Map.entry("kms-asymmetric", this::setupAsymmetric)
        );
    }

    @Override
    public Map<String, TestFn> teardowns() {
        return Map.ofEntries(
                Map.entry("kms-keys",       this::teardownKeys),
                Map.entry("kms-crypto",     ctx -> scheduleDeleteSilently(ctx.getString("kmsCryptoKeyId"))),
                Map.entry("kms-asymmetric", ctx -> scheduleDeleteSilently(ctx.getString("kmsAsymKeyId")))
        );
    }

    // ── kms-keys ───────────────────────────────────────────────────────────────

    private void setupKeys(TestContext ctx) throws Exception {
        var resp = kms().createKey(r -> r.description("compat-test-key").keyUsage(KeyUsageType.ENCRYPT_DECRYPT));
        ctx.set("kmsKeyId", resp.keyMetadata().keyId());
        ctx.set("kmsAliasName", "alias/compat-" + ctx.runId());
    }

    private void teardownKeys(TestContext ctx) {
        // Delete alias first, then schedule key deletion.
        String alias = ctx.getString("kmsAliasName");
        if (alias != null) {
            try { kms().deleteAlias(r -> r.aliasName(alias)); } catch (Exception ignored) {}
        }
        scheduleDeleteSilently(ctx.getString("kmsKeyId"));
    }

    private void createKey(TestContext ctx) {
        Assertions.assertNotBlank(ctx.getString("kmsKeyId"), "kmsKeyId");
    }

    private void describeKey(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        var resp = kms().describeKey(r -> r.keyId(keyId));
        Assertions.assertEquals(keyId, resp.keyMetadata().keyId(), "DescribeKey: keyId mismatch");
        Assertions.assertEquals(KeyState.ENABLED, resp.keyMetadata().keyState(), "DescribeKey: key not ENABLED");
    }

    private void createKmsAlias(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        String alias = ctx.getString("kmsAliasName");
        kms().createAlias(r -> r.aliasName(alias).targetKeyId(keyId));
    }

    private void listKmsAliases(TestContext ctx) throws Exception {
        String alias = ctx.getString("kmsAliasName");
        var resp = kms().listAliases();
        boolean found = resp.aliases().stream().anyMatch(a -> a.aliasName().equals(alias));
        Assertions.assertTrue(found, "ListKmsAliases: alias " + alias + " not found");
    }

    private void listKeys(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        var resp = kms().listKeys();
        boolean found = resp.keys().stream().anyMatch(k -> k.keyId().equals(keyId));
        Assertions.assertTrue(found, "ListKeys: key not found");
    }

    private void disableKey(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        kms().disableKey(r -> r.keyId(keyId));
        var resp = kms().describeKey(r -> r.keyId(keyId));
        Assertions.assertEquals(KeyState.DISABLED, resp.keyMetadata().keyState(), "DisableKey: state should be DISABLED");
    }

    private void enableKey(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        kms().enableKey(r -> r.keyId(keyId));
        var resp = kms().describeKey(r -> r.keyId(keyId));
        Assertions.assertEquals(KeyState.ENABLED, resp.keyMetadata().keyState(), "EnableKey: state should be ENABLED");
    }

    private void scheduleKeyDeletion(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        kms().scheduleKeyDeletion(r -> r.keyId(keyId).pendingWindowInDays(7));
        var resp = kms().describeKey(r -> r.keyId(keyId));
        Assertions.assertEquals(KeyState.PENDING_DELETION, resp.keyMetadata().keyState(),
                "ScheduleKeyDeletion: state should be PENDING_DELETION");
    }

    private void cancelKeyDeletion(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        kms().cancelKeyDeletion(r -> r.keyId(keyId));
        var resp = kms().describeKey(r -> r.keyId(keyId));
        Assertions.assertEquals(KeyState.DISABLED, resp.keyMetadata().keyState(),
                "CancelKeyDeletion: state should be DISABLED after cancel");
    }

    private void tagKmsResource(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        kms().tagResource(r -> r.keyId(keyId).tags(
                Tag.builder().tagKey("env").tagValue("test").build()));
    }

    private void untagKmsResource(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        kms().untagResource(r -> r.keyId(keyId).tagKeys("env"));
    }

    private void listKmsResourceTags(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsKeyId");
        var resp = kms().listResourceTags(r -> r.keyId(keyId));
        Assertions.assertNotNull(resp.tags(), "ListKMSResourceTags: tags is null");
    }

    // ── kms-crypto ────────────────────────────────────────────────────────────

    private void setupCrypto(TestContext ctx) throws Exception {
        var resp = kms().createKey(r -> r.description("compat-crypto-key").keyUsage(KeyUsageType.ENCRYPT_DECRYPT));
        ctx.set("kmsCryptoKeyId", resp.keyMetadata().keyId());
    }

    private void encrypt(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsCryptoKeyId");
        var resp = kms().encrypt(r -> r.keyId(keyId)
                .plaintext(SdkBytes.fromUtf8String("hello KMS")));
        Assertions.assertNotNull(resp.ciphertextBlob(), "Encrypt: ciphertextBlob");
        ctx.set("kmsCiphertext", resp.ciphertextBlob());
    }

    private void decrypt(TestContext ctx) throws Exception {
        String keyId     = ctx.getString("kmsCryptoKeyId");
        SdkBytes cipher  = ctx.get("kmsCiphertext");
        Assertions.assertNotNull(cipher, "Decrypt: no ciphertext from Encrypt step");
        var resp = kms().decrypt(r -> r.keyId(keyId).ciphertextBlob(cipher));
        String plaintext = resp.plaintext().asUtf8String();
        Assertions.assertEquals("hello KMS", plaintext, "Decrypt: plaintext mismatch");
    }

    private void generateDataKey(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsCryptoKeyId");
        var resp = kms().generateDataKey(r -> r.keyId(keyId).keySpec(DataKeySpec.AES_256));
        Assertions.assertNotNull(resp.plaintext(), "GenerateDataKey: plaintext");
        Assertions.assertNotNull(resp.ciphertextBlob(), "GenerateDataKey: ciphertextBlob");
    }

    private void generateDataKeyWithoutPlaintext(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsCryptoKeyId");
        var resp = kms().generateDataKeyWithoutPlaintext(r -> r.keyId(keyId).keySpec(DataKeySpec.AES_256));
        Assertions.assertNotNull(resp.ciphertextBlob(), "GenerateDataKeyWithoutPlaintext: ciphertextBlob");
    }

    // ── kms-asymmetric ────────────────────────────────────────────────────────

    private void setupAsymmetric(TestContext ctx) throws Exception {
        var resp = kms().createKey(r -> r
                .description("compat-asym-key")
                .keyUsage(KeyUsageType.SIGN_VERIFY)
                .keySpec(KeySpec.RSA_2048));
        ctx.set("kmsAsymKeyId", resp.keyMetadata().keyId());
    }

    private void sign(TestContext ctx) throws Exception {
        String keyId = ctx.getString("kmsAsymKeyId");
        var resp = kms().sign(r -> r.keyId(keyId)
                .message(SdkBytes.fromUtf8String("sign this message"))
                .messageType(MessageType.RAW)
                .signingAlgorithm(SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256));
        Assertions.assertNotNull(resp.signature(), "Sign: signature");
        ctx.set("kmsSignature", resp.signature());
    }

    private void verify(TestContext ctx) throws Exception {
        String keyId      = ctx.getString("kmsAsymKeyId");
        SdkBytes sigBytes = ctx.get("kmsSignature");
        Assertions.assertNotNull(sigBytes, "Verify: no signature from Sign step");
        var resp = kms().verify(r -> r.keyId(keyId)
                .message(SdkBytes.fromUtf8String("sign this message"))
                .messageType(MessageType.RAW)
                .signingAlgorithm(SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256)
                .signature(sigBytes));
        Assertions.assertTrue(resp.signatureValid(), "Verify: signature is not valid");
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private void scheduleDeleteSilently(String keyId) {
        if (keyId == null) return;
        try { kms().scheduleKeyDeletion(r -> r.keyId(keyId).pendingWindowInDays(7)); }
        catch (Exception ignored) {}
    }
}

File: compat/suites/dotnet-sdk/Groups/KmsGroup.cs
using System.Text;
using Amazon.KeyManagementService;
using Amazon.KeyManagementService.Model;
using OvercastCompat.Clients;
using OvercastCompat.Harness;

namespace OvercastCompat.Groups;

public sealed class KmsGroup(AwsClients clients) : IServiceGroup
{
    public IReadOnlyDictionary<string, TestFn> Impls() => new Dictionary<string, TestFn>(StringComparer.Ordinal)
    {
        ["CreateKey"] = CreateKeyAsync,
        ["DescribeKey"] = DescribeKeyAsync,
        ["CreateKmsAlias"] = CreateKmsAliasAsync,
        ["ListKmsAliases"] = ListKmsAliasesAsync,
        ["ListKeys"] = ListKeysAsync,
        ["DisableKey"] = DisableKeyAsync,
        ["EnableKey"] = EnableKeyAsync,
        ["ScheduleKeyDeletion"] = ScheduleKeyDeletionAsync,
        ["CancelKeyDeletion"] = CancelKeyDeletionAsync,
        ["TagKMSResource"] = TagKMSResourceAsync,
        ["ListKMSResourceTags"] = ListKMSResourceTagsAsync,
        ["UntagKMSResource"] = UntagKMSResourceAsync,
        ["Encrypt"] = EncryptAsync,
        ["Decrypt"] = DecryptAsync,
        ["GenerateDataKey"] = GenerateDataKeyAsync,
        ["GenerateDataKeyWithoutPlaintext"] = GenerateDataKeyWithoutPlaintextAsync,
        ["Sign"] = SignAsync,
        ["Verify"] = VerifyAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Setups() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["kms-keys"] = SetupKeysAsync,
        ["kms-crypto"] = SetupCryptoAsync,
        ["kms-asymmetric"] = SetupAsymmetricAsync,
    };

    public IReadOnlyDictionary<string, SetupFn> Teardowns() => new Dictionary<string, SetupFn>(StringComparer.Ordinal)
    {
        ["kms-keys"] = TeardownKeysAsync,
        ["kms-crypto"] = TeardownCryptoAsync,
        ["kms-asymmetric"] = TeardownAsymmetricAsync,
    };

    // ── kms-keys ──

    private async Task SetupKeysAsync(TestContext context)
    {
        var response = await clients.KMS().CreateKeyAsync(new CreateKeyRequest());
        context.Set("KmsKeyId", response.KeyMetadata.KeyId);
    }

    private async Task CreateKeyAsync(TestContext context)
    {
        var response = await clients.KMS().CreateKeyAsync(new CreateKeyRequest());
        var keyId = response.KeyMetadata.KeyId;
        Assertions.NotBlank(keyId, "CreateKey: KeyId");
        Assertions.NotBlank(response.KeyMetadata.Arn, "CreateKey: Arn");
        try
        {
            var list = await clients.KMS().ListKeysAsync(new ListKeysRequest());
            Assertions.True(list.Keys.Any(k => k.KeyId == keyId), $"CreateKey: key {keyId} not found in ListKeys (runId={context.RunId})");
        }
        finally
        {
            try { await clients.KMS().ScheduleKeyDeletionAsync(new ScheduleKeyDeletionRequest { KeyId = keyId, PendingWindowInDays = 7 }); } catch { }
        }
    }

    private async Task DescribeKeyAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsKeyId");
        var response = await clients.KMS().DescribeKeyAsync(new DescribeKeyRequest { KeyId = keyId });
        Assertions.Equal(keyId, response.KeyMetadata.KeyId, "DescribeKey: KeyId mismatch");
        Assertions.NotBlank(response.KeyMetadata.Arn, "DescribeKey: Arn");
    }

    private async Task CreateKmsAliasAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsKeyId");
        var aliasName = $"alias/{context.RunId}-compat-key";
        await clients.KMS().CreateAliasAsync(new CreateAliasRequest { AliasName = aliasName, TargetKeyId = keyId });
        try
        {
            var aliases = await clients.KMS().ListAliasesAsync(new ListAliasesRequest { KeyId = keyId });
            Assertions.True(aliases.Aliases.Any(a => a.AliasName == aliasName), $"CreateKmsAlias: alias {aliasName} not found (runId={context.RunId})");
        }
        finally
        {
            try { await clients.KMS().DeleteAliasAsync(new DeleteAliasRequest { AliasName = aliasName }); } catch { }
        }
    }

    private async Task ListKmsAliasesAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsKeyId");
        var aliasName = $"alias/{context.RunId}-list-aliases";
        await clients.KMS().CreateAliasAsync(new CreateAliasRequest { AliasName = aliasName, TargetKeyId = keyId });
        try
        {
            var aliases = await clients.KMS().ListAliasesAsync(new ListAliasesRequest { KeyId = keyId });
            Assertions.True(aliases.Aliases.Any(a => a.AliasName == aliasName), $"ListKmsAliases: alias {aliasName} not found (runId={context.RunId})");
            Assertions.NotBlank(aliases.Aliases[0].AliasName, "ListKmsAliases: AliasName");
        }
        finally
        {
            try { await clients.KMS().DeleteAliasAsync(new DeleteAliasRequest { AliasName = aliasName }); } catch { }
        }
    }

    private async Task ListKeysAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsKeyId");
        var list = await clients.KMS().ListKeysAsync(new ListKeysRequest());
        Assertions.True(list.Keys.Count > 0, "ListKeys: expected non-empty keys");
        Assertions.True(list.Keys.Any(k => k.KeyId == keyId), $"ListKeys: key {keyId} not found (runId={context.RunId})");
    }

    private async Task DisableKeyAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsKeyId");
        await clients.KMS().DisableKeyAsync(new DisableKeyRequest { KeyId = keyId });
        var desc = await clients.KMS().DescribeKeyAsync(new DescribeKeyRequest { KeyId = keyId });
        Assertions.Equal(KeyState.Disabled, desc.KeyMetadata.KeyState, "DisableKey: KeyState");
        await clients.KMS().EnableKeyAsync(new EnableKeyRequest { KeyId = keyId });
    }

    private async Task EnableKeyAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsKeyId");
        await clients.KMS().EnableKeyAsync(new EnableKeyRequest { KeyId = keyId });
        var desc = await clients.KMS().DescribeKeyAsync(new DescribeKeyRequest { KeyId = keyId });
        Assertions.Equal(KeyState.Enabled, desc.KeyMetadata.KeyState, "EnableKey: KeyState should be Enabled");
    }

    private async Task ScheduleKeyDeletionAsync(TestContext context)
    {
        var createResp = await clients.KMS().CreateKeyAsync(new CreateKeyRequest());
        var keyId = createResp.KeyMetadata.KeyId;
        var schedResp = await clients.KMS().ScheduleKeyDeletionAsync(new ScheduleKeyDeletionRequest { KeyId = keyId, PendingWindowInDays = 7 });
        Assertions.NotBlank(schedResp.KeyId, "ScheduleKeyDeletion: KeyId");
        var desc = await clients.KMS().DescribeKeyAsync(new DescribeKeyRequest { KeyId = keyId });
        Assertions.Equal(KeyState.PendingDeletion, desc.KeyMetadata.KeyState, "ScheduleKeyDeletion: KeyState");
    }

    private async Task CancelKeyDeletionAsync(TestContext context)
    {
        var createResp = await clients.KMS().CreateKeyAsync(new CreateKeyRequest());
        var keyId = createResp.KeyMetadata.KeyId;
        try
        {
            await clients.KMS().ScheduleKeyDeletionAsync(new ScheduleKeyDeletionRequest { KeyId = keyId, PendingWindowInDays = 7 });
            await clients.KMS().CancelKeyDeletionAsync(new CancelKeyDeletionRequest { KeyId = keyId });
            var desc = await clients.KMS().DescribeKeyAsync(new DescribeKeyRequest { KeyId = keyId });
            Assertions.Equal(KeyState.Disabled, desc.KeyMetadata.KeyState, "CancelKeyDeletion: KeyState");
        }
        finally
        {
            try { await clients.KMS().ScheduleKeyDeletionAsync(new ScheduleKeyDeletionRequest { KeyId = keyId, PendingWindowInDays = 7 }); } catch { }
        }
    }

    private async Task TagKMSResourceAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsKeyId");
        await clients.KMS().TagResourceAsync(new TagResourceRequest
        {
            KeyId = keyId,
            Tags = [new Tag { TagKey = "project", TagValue = "overcast" }],
        });
        var tags = await clients.KMS().ListResourceTagsAsync(new ListResourceTagsRequest { KeyId = keyId });
        Assertions.True(tags.Tags.Any(t => t.TagKey == "project"), "TagKMSResource: 'project' tag not found");
        Assertions.Equal("overcast", tags.Tags.First(t => t.TagKey == "project").TagValue, "TagKMSResource: tag value mismatch");
    }

    private async Task ListKMSResourceTagsAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsKeyId");
        var tags = await clients.KMS().ListResourceTagsAsync(new ListResourceTagsRequest { KeyId = keyId });
        Assertions.True(tags.Tags.Any(t => t.TagKey == "project"), "ListKMSResourceTags: 'project' tag not found");
    }

    private async Task UntagKMSResourceAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsKeyId");
        await clients.KMS().UntagResourceAsync(new UntagResourceRequest
        {
            KeyId = keyId,
            TagKeys = ["project"],
        });
        var tags = await clients.KMS().ListResourceTagsAsync(new ListResourceTagsRequest { KeyId = keyId });
        Assertions.False(tags.Tags.Any(t => t.TagKey == "project"), "UntagKMSResource: 'project' tag still present");
    }

    private async Task TeardownKeysAsync(TestContext context)
    {
        var keyId = context.GetString("KmsKeyId");
        if (!string.IsNullOrWhiteSpace(keyId))
        {
            try
            {
                var aliases = await clients.KMS().ListAliasesAsync(new ListAliasesRequest { KeyId = keyId });
                foreach (var alias in aliases.Aliases)
                {
                    try { await clients.KMS().DeleteAliasAsync(new DeleteAliasRequest { AliasName = alias.AliasName }); } catch { }
                }
            }
            catch { }
            try { await clients.KMS().ScheduleKeyDeletionAsync(new ScheduleKeyDeletionRequest { KeyId = keyId, PendingWindowInDays = 7 }); } catch { }
        }
    }

    // ── kms-crypto ──

    private async Task SetupCryptoAsync(TestContext context)
    {
        var response = await clients.KMS().CreateKeyAsync(new CreateKeyRequest());
        context.Set("KmsCryptoKeyId", response.KeyMetadata.KeyId);
    }

    private async Task EncryptAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsCryptoKeyId");
        using var plaintext = new MemoryStream(Encoding.UTF8.GetBytes("hello"));
        var response = await clients.KMS().EncryptAsync(new EncryptRequest { KeyId = keyId, Plaintext = plaintext });
        Assertions.GreaterThan(0, response.CiphertextBlob.Length, "Encrypt: CiphertextBlob");
        Assertions.NotBlank(response.KeyId, "Encrypt: KeyId");
    }

    private async Task DecryptAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsCryptoKeyId");
        using var plaintext = new MemoryStream(Encoding.UTF8.GetBytes("hello overcast"));
        var encryptResponse = await clients.KMS().EncryptAsync(new EncryptRequest { KeyId = keyId, Plaintext = plaintext });
        var decryptResponse = await clients.KMS().DecryptAsync(new DecryptRequest
        {
            KeyId = keyId,
            CiphertextBlob = encryptResponse.CiphertextBlob,
        });
        using var reader = new StreamReader(decryptResponse.Plaintext);
        var decrypted = await reader.ReadToEndAsync();
        Assertions.Equal("hello overcast", decrypted, "Decrypt: round trip mismatch");
    }

    private async Task GenerateDataKeyAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsCryptoKeyId");
        var response = await clients.KMS().GenerateDataKeyAsync(new GenerateDataKeyRequest
        {
            KeyId = keyId,
            KeySpec = DataKeySpec.AES_256,
        });
        Assertions.GreaterThan(0, response.CiphertextBlob.Length, "GenerateDataKey: CiphertextBlob");
        Assertions.GreaterThan(0, response.Plaintext.Length, "GenerateDataKey: Plaintext");
        Assertions.NotBlank(response.KeyId, "GenerateDataKey: KeyId");
    }

    private async Task GenerateDataKeyWithoutPlaintextAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsCryptoKeyId");
        var response = await clients.KMS().GenerateDataKeyWithoutPlaintextAsync(new GenerateDataKeyWithoutPlaintextRequest
        {
            KeyId = keyId,
            KeySpec = DataKeySpec.AES_256,
        });
        Assertions.GreaterThan(0, response.CiphertextBlob.Length, "GenerateDataKeyWithoutPlaintext: CiphertextBlob");
        Assertions.NotBlank(response.KeyId, "GenerateDataKeyWithoutPlaintext: KeyId");
    }

    private async Task TeardownCryptoAsync(TestContext context)
    {
        var keyId = context.GetString("KmsCryptoKeyId");
        if (!string.IsNullOrWhiteSpace(keyId))
        {
            try { await clients.KMS().ScheduleKeyDeletionAsync(new ScheduleKeyDeletionRequest { KeyId = keyId, PendingWindowInDays = 7 }); } catch { }
        }
    }

    // ── kms-asymmetric ──

    private async Task SetupAsymmetricAsync(TestContext context)
    {
        var response = await clients.KMS().CreateKeyAsync(new CreateKeyRequest
        {
            KeySpec = "RSA_2048",
            KeyUsage = KeyUsageType.SIGN_VERIFY,
        });
        context.Set("KmsAsymmetricKeyId", response.KeyMetadata.KeyId);
    }

    private async Task SignAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsAsymmetricKeyId");
        using var message = new MemoryStream(Encoding.UTF8.GetBytes("hello"));
        var response = await clients.KMS().SignAsync(new SignRequest
        {
            KeyId = keyId,
            Message = message,
            SigningAlgorithm = SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256,
        });
        Assertions.GreaterThan(0, response.Signature.Length, "Sign: signature");
        Assertions.NotBlank(response.KeyId, "Sign: KeyId");
    }

    private async Task VerifyAsync(TestContext context)
    {
        var keyId = RequireKeyId(context, "KmsAsymmetricKeyId");
        using var message = new MemoryStream(Encoding.UTF8.GetBytes("hello"));
        var signResponse = await clients.KMS().SignAsync(new SignRequest
        {
            KeyId = keyId,
            Message = message,
            SigningAlgorithm = SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256,
        });
        using var verifyMessage = new MemoryStream(Encoding.UTF8.GetBytes("hello"));
        var verifyResponse = await clients.KMS().VerifyAsync(new VerifyRequest
        {
            KeyId = keyId,
            Message = verifyMessage,
            Signature = signResponse.Signature,
            SigningAlgorithm = SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256,
        });
        Assertions.True(verifyResponse.SignatureValid ?? false, "Verify: SignatureValid");
    }

    private async Task TeardownAsymmetricAsync(TestContext context)
    {
        var keyId = context.GetString("KmsAsymmetricKeyId");
        if (!string.IsNullOrWhiteSpace(keyId))
        {
            try { await clients.KMS().ScheduleKeyDeletionAsync(new ScheduleKeyDeletionRequest { KeyId = keyId, PendingWindowInDays = 7 }); } catch { }
        }
    }

    private static string RequireKeyId(TestContext context, string key)
    {
        return context.GetString(key) ?? throw new InvalidOperationException($"{key} not set");
    }
}
